pkg/catalog/catalog.go

```diff
@@ -65,6 +74,105 @@ func GetDevfileRegistries(registryName string) ([]Registry, error) {
 	return devfileRegistries, nil
 }
 
+// ListDevfileComponents lists all the available devfile components
+func (o *CatalogClient) ListDevfileComponents(registryName string) (DevfileComponentTypeList, error) {
+	catalogDevfileList := &DevfileComponentTypeList{}
+	var err error
+
+	// TODO: consider caching registry information for better performance since it should be fairly stable over time
+	// Get devfile registries
+	catalogDevfileList.DevfileRegistries, err = o.GetDevfileRegistries(registryName)
+	if err != nil {
+		return *catalogDevfileList, err
+	}
+	if catalogDevfileList.DevfileRegistries == nil {
+		return *catalogDevfileList, nil
+	}
+
+	// first retrieve the indices for each registry, concurrently
+	devfileIndicesMutex := &sync.Mutex{}
+	retrieveRegistryIndices := util.NewConcurrentTasks(len(catalogDevfileList.DevfileRegistries))
+
+	// The 2D slice index is the priority of the registry (highest priority has highest index)
+	// and the element is the devfile slice that belongs to the registry
+	registrySlice := make([][]DevfileComponentType, len(catalogDevfileList.DevfileRegistries))
+	for regPriority, reg := range catalogDevfileList.DevfileRegistries {
+		// Load the devfile registry index.json
+		registry := reg                 // Needed to prevent the lambda from capturing the value
+		registryPriority := regPriority // Needed to prevent the lambda from capturing the value
+		retrieveRegistryIndices.Add(util.ConcurrentTask{ToRun: func(errChannel chan error) {
+			registryDevfiles, err := getRegistryDevfiles(registry)
+			if err != nil {
+				log.Warningf("Registry %s is not set up properly with error: %v, please check the registry URL and credential (refer `odo registry update --help`)\n", registry.Name, err)
+				return
+			}
+
+			devfileIndicesMutex.Lock()
+			registrySlice[registryPriority] = registryDevfiles
+			devfileIndicesMutex.Unlock()
+		}})
+	}
+	if err := retrieveRegistryIndices.Run(); err != nil {
+		return *catalogDevfileList, err
+	}
+
+	for _, registryDevfiles := range registrySlice {
+		catalogDevfileList.Items = append(catalogDevfileList.Items, registryDevfiles...)
+	}
+
+	return *catalogDevfileList, nil
+}
+
+// GetStarterProjectsNames returns the list of starter projects in a devfile,
+// by temporarily downloading the devile
+func (o *CatalogClient) GetStarterProjectsNames(details DevfileComponentType) ([]string, error) {
+	tmpDir, err := ioutil.TempDir("", "odoinit")
+	if err != nil {
+		return nil, err
+	}
```

Comment:
Yes, you are right, I'll work on adding the fs dependency to the catalog Client

Reply:
I was wondering if we can use in-memory FS instead of storing on disk. I know it's not the kind of optimization we need right now, I was just wondering about its possibility. What you have replaced it with still stores it on disk.

This is not a PR feedback, it's a conversation.